var/spack/repos/builtin/packages/singularity/package.py

```diff
@@ -42,23 +42,24 @@ def gopath(self):
 
     @property
     def sylabs_gopath_dir(self):
-        return join_path(self.gopath, 'src/github.com/sylabs/')
+        return join_path(self.stage.source_path, 'github.com/sylabs/')
 
     @property
     def singularity_gopath_dir(self):
         return join_path(self.sylabs_gopath_dir, 'singularity')
 
-    # Unpack the tarball as usual, then move the src dir into
-    # its home within GOPATH.
+    # Unpack the tarball as usual but ensure it lands in its home within
+    # GOPATH.
     def do_stage(self, mirror_only=False):
         super(Singularity, self).do_stage(mirror_only)
-        source_path = self.stage.source_path
         if not os.path.exists(self.singularity_gopath_dir):
+            go_source_path = join_path(self.stage.path, 'go')
```

Comment:
Since gopath is no longer used (the property) it would be easier to understand this block if (instead of assembling `go_source_path` on the fly, we instead used the updated property `self.gopath`. Thoughts?

Reply:
The property is being used in `setup_environment` for `GOPATH`, which I assumed should stay as the `stage.path`. 

The variable `go_source_path` is a temporary renaming of the fetched/staged files so they can be moved to the expected `GOPATH` location.  This is done because a parent directory -- `stage.source_path` cannot be moved/renamed to a child directory `singularity_gopath_dir`.

Hopefully adding comments to the code will help.  Let me do that now.

Reply:
This builds for me on CentOS7 and w/ the comments it's reasonably clear what's going on.

Reply:
@tldahlgren although I see potential future issues with the GO_PATH, I think your issue being more urgent and the relative few users that would install Singularity aren't enough to stop your progress. As @hartzell mentioned, it does indeed install, so let's deal with any future issues when they come up. Likely we'll update a lot of these golang recipes to build from modules anyway, so all this path nonsense wouldn't matter anyway.

Reply:
Oops, just saw this bit after responding to the earlier comments higher up the thread.  Ignore that bit if you'd like....

Reply:
> build from modules anyway

This is a topic for another thread, but before you get too comfortable, think about this:  Spack frowns on tools that go out on their own and download things from the Internet, no matter how well intentioned they may be.  

For projects that don't vendor their prerequisites, `go` looks at the `go.mod` and `go.sum` files and grabs what it needs as it's building.  Sigh.... [hopefully, singularity will continue to vendor its dependencies even when it starts using modules, that's perfectly legal!]

The proper Spack thing to do is probably going to be to create "`resources`" for each of the things listed in the `go.mod` file (and recursively) and write "resource stanzas" in the `package.py` file.   Then `spack fetch` [SIC?] can go and get them, tuck them into the appropriate dir, and away things will go.

`caddy` has 16 things listed in its `go.mod` but when you `go mod download` them, you end up with 46 things (recursive...).  Other tools have even bigger combinatorial explosions....

Unlike other languages that do this (e.g. lua-rocks), Go goes to great lengths to ensure that the thing it downloads is the thing that it expects.  We'll have a file full of checksums (`go.sum`) *within* the tarball that we checksum ourselves.

Perhaps instead of writing resource definitions, Spack can leverage the existing Go machinery somehow?

I think that Spack-folk will need to kick this around a bit....

Reply:
@vsoch and/or @tldahlgren -- seems like this can also be resolved?

Reply:
Yes, I don't think that I'm able to, so someone else please go ahead.